File: leetcode/first-unique-character-in-a-string.cpp
// https://leetcode.com/problems/first-unique-character-in-a-string/

class Solution {
public:
    int firstUniqChar(string s) {
        int positions[26] = {0};
        
        for (int i = 0; i < s.size(); ++i) {
            char c = s[i] - 'a';
            if (positions[c] == -1) {
                continue;
            } else {
                if (positions[c] == 0) {
                    positions[c] = i + 1;
                } else {
                    positions[c] = -1;
                }
            }
        }
        
        int ans = INT_MAX;
        for (int pos : positions) {
            if (pos <= 0) continue;
            ans = min(ans, pos);
        }
        
        return ans == INT_MAX ? -1 : ans - 1;
    }
};

File: leetcode/first-unique-character-in-a-string.py
# https://leetcode.com/problems/first-unique-character-in-a-string/

def find_first_unique_char(s):
    seen_chars = set()
    unique_chars_to_position = {}

    for idx, c in enumerate(s):
        if c in seen_chars:
            continue
        elif c in unique_chars_to_position:
            del unique_chars_to_position[c]
            seen_chars.add(c)
        else:
            unique_chars_to_position[c] = idx

    if not unique_chars_to_position:
        return -1

    return min(unique_chars_to_position.values())

class Solution(object):
    def firstUniqChar(self, s):
        return find_first_unique_char(s)
